### DISCUSSION TITLE
[BUG] Type of input data when using method fit

### DISCUSSION THREAD

**Discussion by hieutv85** [alan-turing-institute/sktime#4674] [Q&A]
**Describe the bug**

Hi everybody. I faced the problem of not matching input data type when using method fit of sktime. To be more specific, when I used fit method, I got that error "TypeError: X is not of a supported input data type.X must be in a supported mtype format for Panel, found <class 'pandas.core.frame.DataFrame'>Use datatypes.check_is_mtype to check conformance with specifications.". 

Type of the input x, y is DataFrame and Series respectively, shape of input x and y is (1187760, 8) and (1187760,) respectively.

I cannot find a solution to this problem. Please support me to fix this. Thank you very much. I also copied my code here and attached here two pictures of my code and the error. 

https://yourimageshare.com/ib/vPpE9a6CDt
https://yourimageshare.com/ib/nIw8s6fHwS

**To Reproduce**

```python
<Paste your code here>

from sktime.classification.kernel_based import RocketClassifier
rocket = RocketClassifier(num_kernels=2000)
rocket.fit(norm_x_train, norm_y_train)



```

**Expected behavior**

**Additional context**

**Versions**
<details>

</details>

**Comment by hieutv85** [comment]
Hi guys, please help me. 
Please support me @fkiraly. Thank you very much.

**Comment by fkiraly** [comment]
Sorry for the wait, @hieutv85, we were busy with pydata London!

Fortunately, the tutorial we gave there was precisely about what you are asking for, at the start it is explained how to get your data in the right format!

Have a look at notebook 2 here:
https://github.com/sktime/sktime-tutorial-pydata-london-2023
(direct link to notebook: https://github.com/sktime/sktime-tutorial-pydata-london-2023/blob/main/notebooks/02_panel_tasks.ipynb)

you need to get the data in one of the valid formats, that could be `numpy3D` (instance, variable, time), or `pandas.DataFrame` with a multiindex (instance, time) and columns being variable.

**Comment by hieutv85** [comment]
Hi @fkiraly, 

Thank you very much for your kind support. I appreciate this. 
Please allow me to ask another question.

I already convert my input data from dataframe and Series (type of X and y respectively) to  pd-multiindex specification with the method pandas.MultiIndex.from_frame. However I still faced the same error when I used model.fit. 

This is the method: https://pandas.pydata.org/docs/reference/api/pandas.MultiIndex.from_frame.html
This is my code to convert from dataframe and Series to  pd-multiindex:

_norm_x_train = pd.MultiIndex.from_frame(norm_x_train)
norm_y_train = pd.MultiIndex.from_frame(norm_y_train.to_frame())_

I also attached two pictures which show my full code and the error here:
https://ibb.co/wd5T8yP
https://ibb.co/8jyTv6H

 Please support me. Thank you very much.

**Comment by hieutv85** [comment]
My input file is pickle file so I used pickle.load to read data. So the type of input data is DataFrame. 
I searched but there are no parameter return_type of pickle.load (as load data function of sktime) so I can not read input data with the type is pd-multiindex.

**Comment by fkiraly** [comment]
It looks like your `norm_x_train` *is* a `MultiIndex` instead of *having* a `MultiIndex`. For the `pd-multiindex` format, you need a `pd.DataFrame` that *has* a multiindex.

**Reply by hieutv85** [reply]
Hi @fkiraly, 

My input X have the input shape is (1187760, 8). With 1187760 is the number of time lags, and 8 is the number of physiological features. 
Could you please suggest me a suitable way to use it as _numpy3D (instance, variable, time)_, or pandas.DataFrame with a _multiindex (instance, time)_? Thank you very much.  I intent to apply sktime multivariate time series classification methods.

With first option, I think to use it as _numpy3D_ then I should use time window with the number of windows is _instance_. With second option, I have no ideal to how to choose another index to make it become _multiindex_. Please suggest me.

**Reply by fkiraly** [reply]
what are your instances in your data? Or, do you only have a single instance (patient, trial, etc)?

**Reply by hieutv85** [reply]
Hi @fkiraly, 

The input data which I use right now has only one instance since it belong to just one patient. This data recorded physiological value of one patient during an amount of time. The data is 3-axis accelerometer data (x,y,z), skin temperature, respiration data, Electromyogram (EMG), Electrocardiogram (ECG), Galvanic skin response (GRS). The frequency of the data is 70Hz, 1187760 is the number of time lags.

The full data is [WESAD ](https://dl.acm.org/doi/10.1145/3242969.3242985). This is contains 15 patient but I only try to apply sktime with each person first so my input data right now belong to just one patient. I also only used 8/12 physiological features of WESAD. 

If you need anymore information to suggest me which input type of sktime is suitable for my input data then please tell me. 
I am just new with time series analysis. Please support me. Thank you very much.

**Reply by hieutv85** [reply]
The hyperlink which I provide is the paper of the dataset WESAD.

**Reply by pascalamadeus** [reply]
Hi @fkiraly,

could you specify what you mean by "It looks like your norm_x_train is a MultiIndex instead of having a MultiIndex. For the pd-multiindex format, you need a pd.DataFrame that has a multiindex."? I´m not really clear about the difference.

I get the same error message as @hieutv85 using a hierachical Multiindex pd.DataFrame as input data to rocketClassifier. My dataframe is structured as instructed in the docs (https://github.com/sktime/sktime/blob/main/examples/AA_datatypes_and_datasets.ipynb). I got the error even when I used one of the example data sets for hierachical Multiindex df as input data (I´ve created an y_train array as label array). The error occours for X:

```
from sktime.datatypes import get_examples
X_train = get_examples(mtype="pd_multiindex_hier", as_scitype="Hierarchical")[0]
y_train = np.array(['true','false'])

rocket = RocketClassifier(num_kernels=2000)
rocket.fit(X_train , y_train )

TypeError: X is not of a supported input data type.X must be in a supported mtype format for Panel, found <class 'pandas.core.frame.DataFrame'>Use datatypes.check_is_mtype to check conformance with specifications.
```

When I check the conformance using check_is_mtype , I get the following (expected) outcome: 

```
# Check mtype
check_is_mtype(X_train, mtype="pd_multiindex_hier", return_metadata=True)

Out:
(True,
 None,
 {'is_univariate': False,
  'is_empty': False,
  'has_nans': False,
  'n_instances': 6,
  'is_one_series': False,
  'is_equal_length': True,
  'is_equally_spaced': True,
  'n_panels': 2,
  'is_one_panel': False,
  'mtype': 'pd_multiindex_hier',
  'scitype': 'Hierarchical'})
```
